internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		sl.l.Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
+		iplist = nil
+	}
+
+	sl.mlMux.Lock()
+	sl.mlSpeakerIPs = iplist
+	sl.mlMux.Unlock()
+
+	// Update mlSpeakerIPs in the background.
+	go sl.updateSpeakerIPs()
+
+	go sl.memberlistWatchEvents()
+	go sl.joinMembers()
+}
+
+// updateSpeakerIPs runs forever updating the sl.mlSpeakerIPs slice with the
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) updateSpeakerIPs() {
+	for {
+		// This either stops the loop or makes sure to wait at least 5
+		// minutes before making another call to sl.mlSpeakers(), which
+		// queries the API server.
+		select {
+		case <-sl.stopCh:
+			return
+		case <-time.After(5 * time.Minute):
+			// This blocks until the API server responds.
+			iplist, err := sl.mlSpeakers()
+			if err != nil {
+				sl.l.Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
+				continue
+			}
+
+			sl.mlMux.Lock()
+			sl.mlSpeakerIPs = iplist
+			sl.mlMux.Unlock()
+		}
+	}
+}
+
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response from the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context. Newer versions of client-go support using a context
+	// in the call sl.client.PodIPs() is using.
+	//
+	// TODO: When updating client-go, this code can be simplified by using a
+	// context in the following way:
+	// If we use a context, we can specify a timeout and then there is no need to
+	// run mlUpdateSpeaker() in the background. We could then call
+	// this function from mlJoin() with a context (timeout) and use
+	// sl.mlSpeakerIPs on failure. On success, we could update
+	// sl.mlSpeakerIPs with the response and continue to do the join.
+	iplist, err := sl.client.PodIPs(sl.namespace, sl.labels)
+	if err != nil {
+		return nil, err
+	}
+
+	return iplist, nil
+}
+
+func (sl *SpeakerList) joinMembers() {
+	// Every one minute, try to rejoin.
+	// This will join nodes that leave the cluster only for a few seconds every 1
+	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
+	// take a while longer
+	ticker := time.NewTicker(1 * time.Minute)
+
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.mlJoin()
+		case <-sl.mlJoinCh:
+			sl.mlJoin()
+		}
+	}
+}
+
+func (sl *SpeakerList) members() map[string]struct{} {
+	members := make(map[string]struct{})
+
+	for _, node := range sl.ml.Members() {
+		ip := node.Addr.String()
+		members[ip] = struct{}{}
+	}
+
+	return members
+}
+
+// mlJoin joins this memberlist cluster with other speaker pods that are not
+// currently members of this cluster. In other words, it does a memberlist.Join()
+// with the IPs in mlSpeakerIPs that are not members of the cluster.
```

Comment:
```suggestion
// mlJoin joins speaker pods that are not members of this cluster
// to the cluster. It performs a memberlist.Join() with the IPs in
// mlSpeakerIPs that are not members of the cluster.
```

- "Join a cluster with a pod" is IMO confusing both at the English level and the technical level.
- "In other words" doesn't contribute to the meaning of this docstring IMO.
- I find "perform" better in the context of a *process* ("I perform a process" vs. "I do a process").